File: go/storage/mkvs/urkel/commit.go
package urkel

import (
	"context"

	"github.com/oasislabs/oasis-core/go/common"
	"github.com/oasislabs/oasis-core/go/common/crypto/hash"
	db "github.com/oasislabs/oasis-core/go/storage/mkvs/urkel/db/api"
	"github.com/oasislabs/oasis-core/go/storage/mkvs/urkel/node"
	"github.com/oasislabs/oasis-core/go/storage/mkvs/urkel/writelog"
)

// Implements Tree.
func (t *tree) CommitKnown(ctx context.Context, root node.Root) (writelog.WriteLog, error) {
	writeLog, _, err := t.commitWithHooks(ctx, root.Namespace, root.Round, func(rootHash hash.Hash) error {
		if !rootHash.Equal(&root.Hash) {
			return ErrKnownRootMismatch
		}

		return nil
	})
	return writeLog, err
}

// Implements Tree.
func (t *tree) Commit(ctx context.Context, namespace common.Namespace, round uint64) (writelog.WriteLog, hash.Hash, error) {
	return t.commitWithHooks(ctx, namespace, round, nil)
}

func (t *tree) commitWithHooks(
	ctx context.Context,
	namespace common.Namespace,
	round uint64,
	beforeDbCommit func(hash.Hash) error,
) (writelog.WriteLog, hash.Hash, error) {
	t.cache.Lock()
	defer t.cache.Unlock()

	if t.cache.isClosed() {
		return nil, hash.Hash{}, ErrClosed
	}

	oldRoot := t.cache.getSyncRoot()
	if oldRoot.IsEmpty() {
		oldRoot.Namespace = namespace
		oldRoot.Round = round
	}

	batch := t.cache.db.NewBatch(oldRoot, false)
	defer batch.Reset()

	subtree := batch.MaybeStartSubtree(nil, 0, t.cache.pendingRoot)

	rootHash, err := doCommit(ctx, t.cache, batch, subtree, 0, t.cache.pendingRoot, &round)
	if err != nil {
		return nil, hash.Hash{}, err
	}
	if err := subtree.Commit(); err != nil {
		return nil, hash.Hash{}, err
	}

	// Perform pre-commit validation if configured.
	if beforeDbCommit != nil {
		if err := beforeDbCommit(rootHash); err != nil {
			return nil, hash.Hash{}, err
		}
	}

	// Store write log summaries.
	var log writelog.WriteLog
	var logAnns writelog.Annotations
	for _, entry := range t.pendingWriteLog {
		// Skip all entries that do not exist after all the updates and
		// did not exist before.
		if entry.value == nil && !entry.existed {
			continue
		}

		log = append(log, writelog.LogEntry{Key: entry.key, Value: entry.value})
		if len(entry.value) == 0 {
			logAnns = append(logAnns, writelog.LogEntryAnnotation{InsertedNode: nil})
		} else {
			logAnns = append(logAnns, writelog.LogEntryAnnotation{InsertedNode: entry.insertedLeaf})
		}
	}

	root := node.Root{
		Namespace: namespace,
		Round:     round,
		Hash:      rootHash,
	}
	if err := batch.PutWriteLog(log, logAnns); err != nil {
		return nil, hash.Hash{}, err
	}

	// Store removed nodes.
	if err := batch.RemoveNodes(t.pendingRemovedNodes); err != nil {
		return nil, hash.Hash{}, err
	}

	// And finally commit to the database.
	if err := batch.Commit(root); err != nil {
		return nil, hash.Hash{}, err
	}

	t.pendingWriteLog = make(map[string]*pendingEntry)
	t.pendingRemovedNodes = nil
	t.cache.setSyncRoot(root)

	return log, rootHash, nil
}

// doCommit commits all dirty nodes and values into the underlying node
// database. This operation may cause committed nodes and values to be
// evicted from the in-memory cache.
func doCommit(
	ctx context.Context,
	cache *cache,
	batch db.Batch,
	subtree db.Subtree,
	depth node.Depth,
	ptr *node.Pointer,
	round *uint64,
) (h hash.Hash, err error) {
	if ptr == nil {
		h.Empty()
		return
	} else if ptr.Clean {
		if err = subtree.VisitCleanNode(depth, ptr); err != nil {
			return
		}
		h = ptr.Hash
		return
	}

	// Pointer is not clean, we need to perform some hash computations.

	// NOTE: Irreversible cache operations like clearing the dirty flags
	//       and updating node/value cache status must be queued via batch
	//       on-commit hooks as the database operations can fail and this
	//       must not cause the in-memory cache to be corrupted.

	switch n := ptr.Node.(type) {
	case nil:
		// Dead node.
		ptr.Hash.Empty()
	case *node.InternalNode:
		// Internal node.
		if n.Clean {
			panic("urkel: non-clean pointer has clean node")
		}

		// Commit internal leaf (considered to be on the same depth as the internal node).
		if _, err = doCommit(ctx, cache, batch, subtree, depth, n.LeafNode, round); err != nil {
			return
		}

		for _, subNode := range []*node.Pointer{n.Left, n.Right} {
			newSubtree := batch.MaybeStartSubtree(subtree, depth+1, subNode)
			if _, err = doCommit(ctx, cache, batch, newSubtree, depth+1, subNode, round); err != nil {
				return
			}
			if newSubtree != subtree {
				if err = newSubtree.Commit(); err != nil {
					return
				}
			}
		}

		if round != nil {
			n.Round = *round
		}
		n.UpdateHash()

		// Store the node.
		if err = subtree.PutNode(depth, ptr); err != nil {
			return
		}

		batch.OnCommit(func() {
			n.Clean = true
		})
		ptr.Hash = n.Hash
	case *node.LeafNode:
		// Leaf node.
		if n.Clean {
			panic("urkel: non-clean pointer has clean node")
		}

		if round != nil {
			n.Round = *round
		}
		n.UpdateHash()

		// Store the node.
		if err = subtree.PutNode(depth, ptr); err != nil {
			return
		}

		batch.OnCommit(func() {
			n.Clean = true
		})
		ptr.Hash = n.Hash
	}

	batch.OnCommit(func() {
		ptr.Clean = true
		// Make node eligible for eviction.
		cache.commitNode(ptr)
	})
	h = ptr.Hash
	return
}

File: runtime/src/storage/mkvs/urkel/tree/commit.rs
use std::sync::Arc;

use failure::Fallible;
use io_context::Context;

use crate::{
    common::crypto::hash::Hash,
    storage::mkvs::{
        urkel::{cache::*, tree::*},
        LogEntry, WriteLog,
    },
};

impl UrkelTree {
    /// Commit tree updates to the underlying database and return
    /// the write log and new merkle root.
    pub fn commit(&mut self, ctx: Context) -> Fallible<(WriteLog, Hash)> {
        let ctx = ctx.freeze();
        let mut update_list: UpdateList<LRUCache> = UpdateList::new();
        let pending_root = self.cache.borrow().get_pending_root();
        let new_hash = _commit(&ctx, pending_root.clone(), &mut update_list)?;

        update_list.commit(&mut self.cache.borrow_mut());

        let mut log: WriteLog = Vec::new();
        for (_, entry) in self.pending_write_log.iter() {
            // Skip all entries that do not exist after all the updates and
            // did not exist before.
            if entry.value.is_none() && !entry.existed {
                continue;
            }
            log.push(LogEntry {
                key: entry.key.clone(),
                value: entry.value.clone().unwrap_or_default(),
            });
        }
        self.pending_write_log.clear();
        self.cache.borrow_mut().set_sync_root(new_hash);

        Ok((log, new_hash))
    }
}

pub fn _commit<C: Cache>(
    ctx: &Arc<Context>,
    ptr: NodePtrRef,
    update_list: &mut UpdateList<C>,
) -> Fallible<Hash> {
    if ptr.borrow().clean {
        return Ok(ptr.borrow().hash);
    }

    match classify_noderef!(? ptr.borrow().node) {
        NodeKind::None => {
            ptr.borrow_mut().hash = Hash::empty_hash();
        }
        NodeKind::Internal => {
            let some_node_ref = ptr.borrow().get_node();
            if some_node_ref.borrow().is_clean() {
                ptr.borrow_mut().hash = some_node_ref.borrow().get_hash();
            } else {
                let int_left = noderef_as!(some_node_ref, Internal).left.clone();
                let int_right = noderef_as!(some_node_ref, Internal).right.clone();

                _commit(ctx, int_left.clone(), update_list)?;
                _commit(ctx, int_right.clone(), update_list)?;

                some_node_ref.borrow_mut().update_hash();
                ptr.borrow_mut().hash = some_node_ref.borrow().get_hash();

                let closure_node_ref = some_node_ref.clone();
                update_list.push(Box::new(move |_| {
                    noderef_as_mut!(closure_node_ref, Internal).clean = true
                }));
            }
        }
        NodeKind::Leaf => {
            let node_ref = ptr.borrow().get_node();
            if node_ref.borrow().is_clean() {
                ptr.borrow_mut().hash = node_ref.borrow().get_hash();
            } else {
                if !noderef_as!(node_ref, Leaf).value.borrow().clean {
                    noderef_as!(node_ref, Leaf).value.borrow_mut().update_hash();

                    let closure_value = noderef_as!(node_ref, Leaf).value.clone();
                    update_list.push(Box::new(move |cache| {
                        closure_value.borrow_mut().clean = true;
                        cache.commit_value(closure_value.clone());
                    }));
                }

                node_ref.borrow_mut().update_hash();
                ptr.borrow_mut().hash = node_ref.borrow().get_hash();

                let closure_node_ref = node_ref.clone();
                update_list.push(Box::new(move |_| {
                    noderef_as_mut!(closure_node_ref, Leaf).clean = true
                }));
            }
        }
    };

    let closure_ptr = ptr.clone();
    update_list.push(Box::new(move |cache| {
        closure_ptr.borrow_mut().clean = true;
        cache.commit_node(closure_ptr.clone());
    }));

    Ok(ptr.borrow().hash)
}
